### ISSUE TITLE
Questions and opinions on the the most recent MinHash and SetSketch

### ISSUE THREAD

**Issue by jianshu93** [dynatrace-research/set-sketch-paper#4]
Hello Otmar,

I am writing to ask your opinion on the most recent MinHash, called circulant permutation MinHash as published here (https://proceedings.mlr.press/v162/li22m.html) since when we were implementing the Joint MLE estimator of SetSketch, due to the fact that the true cardinalities of set is not known (e.g., in our case, genomes, vary quite a lot), we have to use the cardinality estimator first (equation (12)) and then use it for Joint MLE (section 3.2) via the Brent’s method (we were using the golden-section search (https://en.wikipedia.org/wiki/Golden-section_search) since it is strictly concave within an interval). Our implementation and my understanding agreed that the variance of the Joint MLE estimator is limited by the RMSE of cardinality estimation (e.g., approximating sqrt(1.079/m) or worse), despite being much more space efficient. We care about small Jaccard like those around 0.01 or smaller because in genome analysis this is important range where other methods will just be too expensive. For large Jaccard, there is no doubt that Setsketch is both space efficient and accurate enough. For small Jaccard, in the end, we went to densified Minhash (the optimal densification paper, Shrivastava 2017, strictly sqrt(J*(1-J)/m)) and figured out that there is an even better one as mentioned in the beginning, strictly smaller than sqrt(J*(1-J)/m) while being computationally as efficient as Shrivastava 2017 as claimed in the paper. It could be very useful/accurate for small Jaccard. However, the second permutation, to randomly choose hash values from non-empty bins to empty bins cannot be avoided (e.g., must use the original permutation based processing as in the Broder 1998 paper, not hashing based method to approximate permutation), which means the permutation vector, can be very long in practice, e.g., 2^16 to 2^18 et.al., depending on the bucket size K when data dimension is very large, e.g., 2^30. So not so memory efficient, but only one such permutation is needed, and we need to circular shift the permutation vector K times (K is the number of buckets as in one permutation hashing), one after another, so pervious permutation can be released from memory after the shifted permutation is computed. Still, not sure how this will greatly affect computational efficiency because I tested that the circular shift of such a large permutation vector is also very expensive. Just want to hear your opinions on the circular permutation method since you are more familiar with computational optimization of those probabilistic structures. Sorry for the long description. Let me know if I am not clear. 

Thank you very much,

Best,

Jianshu

**Comment by oertl** [comment]
Hi Jianshu,

looking at Fig. 3 of the C-MinHash paper, the variance seems to be improved
significantly only for cardinalities D that do not exceed the size of the
minhash K. Would be interesting to compare those results with that of my
SuperMinHash algorithm https://arxiv.org/pdf/1706.05698.pdf. I would be
surprised if C-MinHash would beat SuperMinHash in terms of variance. Maybe
the variance of both algorithms is very close. The SuperMinHash algorithm
is described in the paper as a streaming algorithm. However, if multiple
passes over the data are possible, one can significantly improve the
performance (see this implementation
https://github.com/dynatrace-oss/hash4j/blob/main/src/main/java/com/dynatrace/hash4j/similarity/SuperMinHashPolicy_v1.java).
For large cardinalities the branch at
https://github.com/dynatrace-oss/hash4j/blob/827420a2e952600d60f443c703ab563d682a6ed3/src/main/java/com/dynatrace/hash4j/similarity/SuperMinHashPolicy_v1.java#L115C41-L115C41
is never executed and the runtime is then similar to One Permutation
Hashing.

Best

Otmar

**Comment by jianshu93** [comment]
Hello Otmar,

We've already implemented the 2 algorithms, superminhash 1 and 2 as described in the paper. We have the problem as described by Jean (Jean implemented this) before: "For moderate m , the expected cost is well below mn. After sketching some thousand of kmer (element of a genome set) we have inner loop of size in the range 1-3. In this case I can repeatedly make tests on my laptop on this file. But as m increase 5000 or 1000 the size of the inner loop cease to decrease and stay above some thousands, which give thousands of inner loop by item sketched! If n is small this is not a problem, but with so large n , even some sequences with the bad case behavior give a catastrophe". We cannot have multiple pass of the data because genomic data are very large (e.g., total close to a million files, each 10^7 elements, 3% to 5% of the elements of each files is duplicated more than once, 2-5 times on average) and if we do so, it will be very expensive (you've already made the suggestion of this one last time, so we cannot improve performance with just one pass). It seems the branch never executed is a good fit, since we have large cardinalities and the minhash K is always much smaller than cardinalities in our case to achieve the accuracy we need. This one is as fast as one permutation hashing but also much more accurate than classic MinHash right, because we want both the performance and accuracy as in the circulated permutation.

Thank you,

Jianshu

**Comment by jianshu93** [comment]
Hello Otmar,
In practice, we need very large sketch size to accurate,e.g., m=12,000 or more. The problem is that the exit of the main loop in sketching is random and we exit from the loop with decreasing probability while size of sketch is increasing so that with large sketch size we fall back to original minhash complexity. (your commentary in java sources). We have them in our implementation but it is just too slow in practice, at least 10 times slower than densified Minhash with similar accuracy because we can use slightly larger sketch size in densified MinHash without losing speed due to O(n + m), where n is at least 100 times larger than m in our case. I feel we do not have a way out due to this reason for speed, but accuracy is better than classic MinHash. 

Thanks,

Jianshu

**Comment by oertl** [comment]
Hi Jianshu,

m >= 12000 is really huge. To better understand the problem: What is the
maximum input size n? I assume that the input is not deduplicated. What is
the maximum number of input elements n'? Is n' in the order of n or much
smaller than n?
I would be interested, if n' is small enough such that 8 x n' bytes fit
into memory. The reason why I ask is that it is not necessary to have
multiple passes over the raw data. It is sufficient to keep the 64-bit hash
values of the elements in memory. If you deduplicate them using the
strategy used by
https://github.com/dynatrace-oss/hash4j/blob/main/src/main/java/com/dynatrace/hash4j/similarity/DistinctElementHashProvider.java,
the space requirements will be in the order of n'.

Best
Otmar

**Comment by jianshu93** [comment]
Hello Otmar,

n is on average 10^7 for each file (number of element), n' (after deduplication) is about 0.95*n. We can deduplicate it but we also have probminhash 3a implemented, for which we cannot deduplicate. So we have to ran an additional step to deduplicate for superminhash right (otherwise it will be too slow). Such large m is used because we want enough accuracy at Jaccard index 0.01 or 0.005, a small m is not good for this range of Jaccard. 

Thanks,

Jianshu

**Comment by oertl** [comment]
Hi Jianshu!

If n is 10^7, it is possible to calculate all 64-bit hash values of all elements in the set in a first step. Keeping those hash values requires 80MB of memory which should be fine. This list of hash values then serves as input for the SuperMinHash algorithm. I tested the SuperMinHash implementation of [hash4j ](https://github.com/dynatrace-oss/hash4j) for n = n' = 10^7 and m=10^5. I did 100 test runs and none of them executed [this branch](https://github.com/dynatrace-oss/hash4j/blob/985936d1c60e2e7a06ef910c6d4f6402bd0fd06d/src/main/java/com/dynatrace/hash4j/similarity/SuperMinHashPolicy_v1.java#L115-L151). So I don't see why SuperMinHash cannot be applied. Deduplication of the element hashes has no significant benefit if there are just 5% duplicates.

Best
Otmar
